ibeckermayer/teleport-interview
# Session management part 1

# Session management part 1
This PR adds code for the database layer and the beginnings of the session management system on the `backend/`, and gives the `frontend/` a global state management system.

## `backend/`
I chose [`sqlx`](https://github.com/jmoiron/sqlx) as a `database/sql` wrapper because it appears to provide better developer ergonomics for marshaling data to and from the database and `structs`.

#### `package model`
`model` is the simplest new package, currently containing just a variable with the SQL statement for creating the account table and the corresponding `Account` struct. Initially `Account` was in the `database` package, but I later realized that I needed it in more than one place (namely, for the session manager in the `auth` package) and so moved it to it's own `model` package in order to solve cyclic import errors.

#### `package database`
`database` is naturally the package for initializing and interacting with the database. Right now it comes hardcoded to use `sqlite3` and I didn't focus at all on making it theoretically compatible with other RDBMS engines. `database.go` contains config and initialization, and `account.go` contains functions for interacting with the account table.

#### `package auth`
`auth` right now is something of an eclectic collection of functions and objects related to authentication. `genkey.go` has utility functions for generating cryptographically random session ID's (and in the future, API keys). `password.go` has the password hashing and hash-checking functions.

The meat of the package is in `sessionmanager.go`, which contains the beginning of an in-memory session manager. The idea is that there will be a single global `*SessionManager`, which will house all of the `Session`s in the `SessionManager.store` map (indexed by `SessionID`). 

In the future (next PR), auth protected endpoints will be sent the `SessionID` in the header, and the `SessionManager` will check that a valid session exists (using appropriate synchronization mechanisms) and then return a copy of the `Session` object. The handlers can then modify/use the `Session` objects as necessary, and then later if the `Session` object needs to be updated, the `SessionManager` can replace the corresponding value in the `store` (again using appropriate synchronization mechanisms).

## `frontend/`
The meat of the changes on the frontend are in `store.jsx`, where I set up a `React.Context.Provider` that is wrapped around the entire `<App />` to make a global `store` state available to any component that consumes it. I then modified `Login.jsx` to update the global store on login and switch to the `/dashboard` route when the user is authenticated.

This design is based on the description in this blog post [here](https://codeburst.io/global-state-with-react-hooks-and-context-api-87019cc4f2cf), with some added code to make use of `localStorage`.

(@alex-kovoy I haven't forgotten about [this](https://github.com/ibeckermayer/teleport-interview/pull/2#discussion_r552391932), planning to upgrade to that in the next PR)

# To Test
In order to realize the changes to the `backend/dev.Dockerfile`, you may need to rebuild the images by running `docker-compose -f docker-compose-dev.yml build --no-cache` (just FYI, since this was causing me major headaches for a while). Then run
```bash
docker-compose -f docker-compose-dev.yml up
```
and navigate to https://0.0.0.0:8080

If you try and login with any credentials other than username/pwd == `dev@goteleport.com`, you should receive an error. If you log in with the correct credentials, you should be routed to `/dashboard`. If you navigate to the app again in another tab, you should get routed quickly to `/dashboard` since the global store should be found in local storage.

You can visually confirm that the `SessionManager` is doing something that seems about right by adding
```go
log.Printf("%+v", lh.sm)
```
to line 72 of `backend/internal/handlers/login.go`
